File: claude-agent-sdk/lib/agent.ts
import Anthropic from "@anthropic-ai/sdk";
import { writeFileSync, readFileSync, existsSync } from "fs";
import { join } from "path";

const client = new Anthropic();

const TAVILY_API_KEY = process.env.TAVILY_API_KEY || "";

const tools: Anthropic.Tool[] = [
  {
    name: "web_search",
    description:
      "Search the web for information about FIFA/EA FC games and revenue data using Tavily API",
    input_schema: {
      type: "object" as const,
      properties: {
        query: {
          type: "string",
          description: "Search query",
        },
      },
      required: ["query"],
    },
  },
  {
    name: "write_output",
    description: "Write the final research output to a markdown file",
    input_schema: {
      type: "object" as const,
      properties: {
        content: {
          type: "string",
          description: "The markdown content with embedded JSON to write",
        },
        filename: {
          type: "string",
          description: "The output filename",
        },
      },
      required: ["content", "filename"],
    },
  },
];

async function tavilySearch(query: string): Promise<string> {
  if (!TAVILY_API_KEY) {
    return JSON.stringify({
      error: "TAVILY_API_KEY not set",
      results: [],
    });
  }

  try {
    const response = await fetch("https://api.tavily.com/search", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        api_key: TAVILY_API_KEY,
        query,
        max_results: 10,
        include_answer: true,
      }),
    });

    if (!response.ok) {
      return JSON.stringify({ error: `Tavily API error: ${response.status}` });
    }

    const data = await response.json();
    return JSON.stringify(data);
  } catch (error) {
    return JSON.stringify({ error: `Search failed: ${error}` });
  }
}

function handleToolCall(
  toolName: string,
  toolInput: Record<string, string>
): Promise<string> | string {
  if (toolName === "web_search") {
    return tavilySearch(toolInput.query);
  }

  if (toolName === "write_output") {
    const projectRoot = join(process.cwd(), "..");
    const outputPath = join(projectRoot, "claude-agent-sdk", toolInput.filename);
    writeFileSync(outputPath, toolInput.content, "utf-8");
    return `Output written to ${outputPath}`;
  }

  return `Unknown tool: ${toolName}`;
}

export async function runResearchAgent(iteration: number): Promise<{
  success: boolean;
  outputFile: string;
  summary: string;
}> {
  const projectRoot = join(process.cwd(), "..");
  const outputFilename = `research-${iteration}-output.md`;
  const outputPath = join(projectRoot, "claude-agent-sdk", outputFilename);

  // Read evolution hints if they exist
  let evolutionHints = "";
  const hintsPath = join(projectRoot, "research", "sdk-evolution-hints.md");
  if (existsSync(hintsPath)) {
    evolutionHints = readFileSync(hintsPath, "utf-8");
  }

  const systemPrompt = `You are a FIFA/EA FC Research Agent (SDK version). Your task is to research ALL FIFA and EA Sports FC video game releases from the 1990s through EA Sports FC 26.

For each game, collect:
- name: Official game title
- year: Release year
- platform: Platform(s)
- publisher: Publisher
- estimatedRevenue: Estimated total revenue in USD
- revenueSource: Where the data came from
- confidence: "low", "medium", or "high"

Include main series, FIFA Street, FIFA Online, FIFA Mobile, and all official titles.

Your output MUST be a markdown file with a summary section followed by a JSON code block with this schema:
{
  "games": [{ "name", "year", "platform", "publisher", "estimatedRevenue", "revenueSource", "confidence" }],
  "totalGamesFound": number,
  "searchQueries": string[],
  "sources": string[]
}

Use the web_search tool to find data. When done, use write_output to save your research.
The output filename must be: ${outputFilename}

${evolutionHints ? `## Previous Iteration Findings\n${evolutionHints}` : ""}`;

  const messages: Anthropic.MessageParam[] = [
    {
      role: "user",
      content: `Research iteration ${iteration}: Find all FIFA/EA FC game releases and revenue data. Use web_search to gather data, then use write_output to save your complete research as "${outputFilename}".`,
    },
  ];

  let response = await client.messages.create({
    model: "claude-sonnet-4-20250514",
    max_tokens: 8192,
    system: systemPrompt,
    tools,
    messages,
  });

  // Agentic loop — keep processing tool calls until done
  while (response.stop_reason === "tool_use") {
    const toolUseBlocks = response.content.filter(
      (block): block is Anthropic.ContentBlock & { type: "tool_use" } =>
        block.type === "tool_use"
    );

    const toolResults: Anthropic.MessageParam = {
      role: "user",
      content: await Promise.all(
        toolUseBlocks.map(async (toolUse) => ({
          type: "tool_result" as const,
          tool_use_id: toolUse.id,
          content: await handleToolCall(
            toolUse.name,
            toolUse.input as Record<string, string>
          ),
        }))
      ),
    };

    messages.push(
      { role: "assistant", content: response.content },
      toolResults
    );

    response = await client.messages.create({
      model: "claude-sonnet-4-20250514",
      max_tokens: 8192,
      system: systemPrompt,
      tools,
      messages,
    });
  }

  // Extract text summary
  const textBlocks = response.content.filter(
    (block): block is Anthropic.TextBlock => block.type === "text"
  );
  const summary = textBlocks.map((b) => b.text).join("\n");

  const success = existsSync(outputPath);

  return {
    success,
    outputFile: outputFilename,
    summary: success
      ? `Research complete. Output written to ${outputFilename}`
      : `Agent finished but output file not found. Agent response: ${summary.slice(0, 500)}`,
  };
}

File: claude-agent-sdk/agent.py
import asyncio
import json
from pathlib import Path
from claude_agent_sdk import query, ClaudeAgentOptions, AssistantMessage, ResultMessage, TextBlock

PROJECT_ROOT = Path(__file__).parent.parent


def load_problem() -> str:
    """Load the research problem definition from problem-statement/problem-statement.md."""
    problem_path = PROJECT_ROOT / "problem" / "problem.md"
    if problem_path.exists():
        return problem_path.read_text(encoding="utf-8")
    return "No problem definition found. Check problem-statement/problem-statement.md."


def load_agent_instructions() -> str:
    """Load the same agent instructions used by CLI (claude-code-cli-games-revenue-researcher.md)."""
    agent_path = PROJECT_ROOT / ".claude" / "agents" / "claude-code-cli-games-revenue-researcher.md"
    if agent_path.exists():
        return agent_path.read_text(encoding="utf-8")
    return ""


def load_evolution_log() -> str:
    """Load SDK evolution log if it exists."""
    log_path = PROJECT_ROOT / "research" / "sdk-evolution-log.md"
    if log_path.exists():
        return log_path.read_text(encoding="utf-8")
    return ""


def load_mcp_config() -> dict:
    """Load MCP server config from .mcp.json."""
    mcp_path = PROJECT_ROOT / ".mcp.json"
    if mcp_path.exists():
        data = json.loads(mcp_path.read_text(encoding="utf-8"))
        return data.get("mcpServers", {})
    return {}


async def _run_research(iteration: int) -> dict:
    """Run the research agent using Claude Agent SDK (uses Max subscription)."""
    problem = load_problem()
    agent_instructions = load_agent_instructions()
    evolution_log = load_evolution_log()
    mcp_servers = load_mcp_config()

    # Ensure output directory exists
    output_dir = PROJECT_ROOT / "research" / f"research-{iteration}" / "claude-agent-sdk"
    output_dir.mkdir(parents=True, exist_ok=True)

    evolution_section = ""
    if evolution_log:
        evolution_section = f"\n\n## SDK Evolution Log (learnings from previous iterations)\n{evolution_log}"

    system_prompt = f"""You are a Research Agent (SDK version). You must produce the SAME output as the CLI Research Agent would produce. Use the same research approach and methodology.

## Research Problem

{problem}

## Agent Instructions (same as CLI agent)

{agent_instructions}

## Output Instructions

You MUST write TWO files:

1. First write `research/research-{iteration}/claude-agent-sdk/reddit-data-{iteration}.md` — Save ALL raw Reddit posts/data you collected. Include post titles, content snippets, scores, and subreddits.

2. Then write `research/research-{iteration}/claude-agent-sdk/research-{iteration}.md` — Your synthesized research with a markdown summary followed by a JSON code block matching the output schema from the research problem.

Write reddit-data FIRST, then the research file.{evolution_section}"""

    prompt = f"Research iteration {iteration}: Execute the research problem. Use the Reddit MCP tools to find data. Produce output matching the CLI agent format exactly. Write reddit-data-{iteration}.md first, then research-{iteration}.md to research/research-{iteration}/claude-agent-sdk/."

    options = ClaudeAgentOptions(
        system_prompt=system_prompt,
        allowed_tools=["Read", "Write", "Edit", "Bash", "Glob", "Grep", "mcp__reddit-mcp-server"],
        permission_mode="dangerouslySkipPermissions",
        cwd=str(PROJECT_ROOT),
        mcp_servers=mcp_servers,
        max_turns=30,
    )

    final_text = ""
    cost = 0.0

    async for message in query(prompt=prompt, options=options):
        if isinstance(message, AssistantMessage):
            for block in message.content:
                if isinstance(block, TextBlock):
                    final_text += block.text
        elif isinstance(message, ResultMessage):
            cost = getattr(message, "total_cost_usd", 0.0) or 0.0

    research_file = output_dir / f"research-{iteration}.md"
    reddit_file = output_dir / f"reddit-data-{iteration}.md"

    return {
        "success": research_file.exists(),
        "outputFile": str(research_file.relative_to(PROJECT_ROOT)) if research_file.exists() else "",
        "redditDataFile": str(reddit_file.relative_to(PROJECT_ROOT)) if reddit_file.exists() else "",
        "summary": f"Research iteration {iteration} complete (cost: ${cost:.4f})"
        if research_file.exists()
        else "Output file not created by agent",
    }


def run_research_agent(iteration: int) -> dict:
    """Sync wrapper for the async research agent."""
    return asyncio.run(_run_research(iteration))
